### ISSUE TITLE
Support more granular permissions than database owner for postgresuser

### ISSUE THREAD

**Issue by arnarg** [movetokube/postgres-operator#18]
Currently PostgresUser can only be a database owner over one database.

It would be better if we could also give more granular permissions to users. It could look something like this:

```
apiVersion: db.movetokube.com/v1alpha1
kind: PostgresUser
metadata:
  name: my-db-user
  namespace: app
spec:
  role: username
  permissions:
  - priv: CREATE
    database: my-db
  - priv: SELECT
    database: other-db
    schema: customer
    table: *
  secretName: my-credential-secret
```

It could also look similar to Ansible's [postgresql_user module](https://docs.ansible.com/ansible/latest/modules/postgresql_user_module.html#examples).

The owner could still be supported by instead of specifying a permissions list we simply specify a database (the current PostgresUser CRD), have a special `priv: OWNER` that we would handle specially by adding the user to the group role that owns the database or simply move the current `PostgresUser` to another CRD called `PostgresOwner` or something similar.

What do you think?

**Comment by arnarg** [comment]
@hitman99 Do you have an opinion on this?

**Comment by hitman99** [comment]
@arnarg I like the idea of permissions, but I'm not a fan of a user having access to multiple databases. It's a security concern - if the credentials are leaked, the blast radius is not limited to only one database. User could just create another PostgresUser CR with different database and gain access to it that way. Do you see any downsides to it?

**Comment by arnarg** [comment]
@hitman99 I agree, that makes more sense.

Something like this?

```
apiVersion: db.movetokube.com/v1alpha1
kind: PostgresUser
metadata:
  name: my-db-user
  namespace: app
spec:
  role: username
  database: my-db
  privileges:
  - priv: CREATE
  - priv: SELECT
    schema: customer
    table: *
  secretName: my-credential-secret
```

`privileges` could also be a comma separated list in a string but I don't know how we would control schema and tables. If privileges is unset it would just be an owner if we want to not break the API.

**Comment by hitman99** [comment]
Yeah, this looks better. You have a good point there, current logic just adds a user to master role (generated or provided by the user) and thus user has full priviledges to the database. How would these granular permissions work with that logic? As after adding user to a group, it already has full access

**Comment by arnarg** [comment]
I'm thinking if you specify a more granular permission you're not added to that group. And I also mentioned in the original issue that we could have another special CRD for database owner, but that would break the API of course.

**Comment by hitman99** [comment]
This might be feasible, but the transfer of the ownership will be very difficult to implement with these new permissions. Currently the user does not own anything, everything is owned by the group role. IF we do not add user to that role, everything that the user has created need to be transfered to the master role before user could be deleted. This sounds simple, but it's a PITA 😄 And the worst thing is - user cannot be deleted in a "_cascading_" mode, all objects that the user owns needs to be removed or transfered first.

**Comment by arnarg** [comment]
That's a good point I hadn't thought about. Will `REASSIGN OWNED BY` be possible?

Lookup who's the owner of the database and reassign objects to that user.

**Comment by hitman99** [comment]
That might work, does that work for schemas, sequences too?

**Comment by arnarg** [comment]
I'm honestly not sure. From the documentation I feel like it reads as all objects but we'd need to test it.

I just want to know what we'd want the CRD to look like before starting to put any actual work into this.

**Comment by arnarg** [comment]
@hitman99 I just tried it and `REASSIGN OWNED BY` does transfer ownership of schema and sequences.

**Comment by hitman99** [comment]
Ok, then it's not as complicated as I thought. As for CR shape/options, we need to decide on the several things:

- what permissions to omplement
- will permissions declared in CR will be directly mapped to Postgres privs
- will there be an abstration layer for CR permissions when mapped to Postgres privs, i.e. `READ` permission would mean read-only access, and `WRITE` permissions would mean you can do anything.

Do you have a particular use case in mind? I think that implementing features for the sake of implementation will not lead to anything usefull

**Comment by arnarg** [comment]
We want to be able to create application roles. So probably `SELECT`, `INSERT`, `UPDATE` and `DELETE` on all tables in schema and/or database.

I think an abstraction layer makes sense to begin with at least.

So `READ` could be just `SELECT` on all tables in schema (if specified, otherwise all schemas in database) and `WRITE` all four above. There isn't even any need for `REASSIGN OWNED BY`, now that I think about it, as there's no ownership on rows in a table.

So you'd use the database owner for migration and creation and the application can just create/mutate/read data in tables. Does this make sense?

**Comment by hitman99** [comment]
I see, it makes more sense to me now. I see only one potential issue with this setup - `REASSIGN OWNED BY` is not needed with a condition that no new resources will be created by the user with these granular permissions. Meaning that the user can only have permissions to mutate existing resources, but should not have any permissions to create new resources like tables, schemas, sequences or indexes.

**Comment by arnarg** [comment]
If you run the following commands you have a user `appuser` that has permissions to insert,delete,update and select on all tables in a schema, shouldn't that be ok?

```
CREATE ROLE appuser WITH LOGIN PASSWORD 'asdf1234';
GRANT USAGE ON SCHEMA customers TO appuser;
GRANT SELECT,INSERT,DELETE,UPDATE ON ALL TABLES IN SCHEMA customers TO appuser;
```

This user will only be able to mutate/delete/insert rows in an existing table. Right?

**Comment by hitman99** [comment]
Yeah, that's correct, but then the CR changes you proposed are missing additional `schema` property, as you grant ussage on schema. But all in all, this new feature looks nice :)

**Comment by arnarg** [comment]
There's one snag I'm seeing here. When I try to `GRANT USAGE` on non-existing schema it errors. So we would probably want to either support creating schema in `Postgres` or `PostgresRole` would need to retry until the schema has been created by someone.

**Comment by arnarg** [comment]
Or even that `PostgresRole` would just create the schema the user wants to get permissions to.

> but then the CR changes you proposed are missing additional schema property

Yeah, I would want this to be a list probably.

**Comment by hitman99** [comment]
> Or even that PostgresRole would just create the schema the user wants to get permissions to.

Correct me if I'm wrong, but if the schema is created by the operator, the `GRANT SELECT,INSERT,DELETE,UPDATE ON ALL TABLES IN SCHEMA customers TO appuser;` will not work as it does not automatically grant privileges upfront, privs are only granted for existing objects.

**Comment by arnarg** [comment]
I meant that we would do `CREATE SCHEMA` just in case and just ignore `duplicate_object` error before granting the user `USAGE` on the schema.

This however seems a bit hacky to me. Maybe it would be better to add a `schemas` list to `Postgres` CRD that would populate the database with schemas.

**Comment by arnarg** [comment]
`CREATE SCHEMA IF NOT EXISTS` is also possible.

**Comment by arnarg** [comment]
@hitman99 So, it turns out that plan will not work. If you `GRANT ALL ON ALL TABLES IN SCHEMA customers TO appuser;` you are only giving `appuser` all privileges to all tables that currently exist in schema `customers`, he will not have any privileges on tables created after that statement is run.

I propose we have three group roles that are created with `Postgres`.
1. The existing owner group role.
2. A reader role. This one can SELECT on all tables in that database (but limited to schemas the operator is aware of).
3. A writer role. Same but has SELECT, INSERT, DELETE and UPDATE.

We then assign new `PostgresUser` to any of those groups depending on `Privileges` property. This will also simplify dropping logic of `PostgresUser`.

This will require postgres-operator to be authoritative over schemas in the database, this will be in a `schemas` list in spec. The database owner group will still be able to create new schemas but the reader and writer roles will not have access to them. This is all due to PostgreSQL short-comings (as far as I can tell).

So on `Postgres` creation the following will happen:

```sql
CREATE DATABASE "dbname";
CREATE ROLE "dbname-group";
ALTER DATABASE "dbname" OWNER TO "dbname-group";
--- The above already happens
--- Create all schemas
CREATE SCHEMA IF NOT EXISTS schemas[i];
CREATE ROLE "dbname-reader";
--- The following needs to happen to all schemas postgres-operator is aware of
GRANT USAGE ON SCHEMA schemas[i] TO "dbname-reader";
ALTER DEFAULT PRIVILEGES FOR ROLE "dbname-group" IN SCHEMA schemas[i] GRANT SELECT ON TABLES TO "dbname-reader";
--- Then the same for dbname-writer except with SELECT,INSERT,DELETE,UPDATE
```

So the CRDs will look like following:

```yaml
apiVersion: db.movetokube.com/v1alpha1
kind: Postgres
metadata:
  name: my-db
spec:
  database: dbname
  schemas:
  - customers
  - stores
```

```yaml
apiVersion: db.movetokube.com/v1alpha1
kind: PostgresUser
metadata:
  name: my-db-user
spec:
  role: username
  database: my-db
  secretName: my-credential-secret
  privileges: OWNER|READ|WRITE
```

PostgreSQL privileges are really fucking complicated 😛 

What do you think?

**Comment by hitman99** [comment]
@arnarg looks ok. And yes, postgres porovileges are quite complicated 😄

**Comment by arnarg** [comment]
This was closed with #19
